test/jdk/java/nio/file/Files/CopyMoveVariations.java

```diff
@@ -0,0 +1,234 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class CopyMoveVariations {
+    enum OpType {
+        COPY,
+        MOVE
+    }
+
+    enum PathType {
+        FILE,
+        DIR,
+        LINK
+    }
+
+    private static Stream<Arguments> params() {
+        List<Arguments> list = new ArrayList<Arguments>();
+
+        boolean[] falseAndTrue = new boolean[] {false, true};
+        for (PathType type : PathType.values()) {
+            String[] modes = new String[] {
+                "---------", "r--r--r--", "-w--w--w-", "rw-rw-rw-"
+            };
+            for (String mode : modes) {
+                for (boolean replaceExisting : falseAndTrue) {
+                    for (boolean targetExists : falseAndTrue) {
+                        list.add(Arguments.of(type, mode, replaceExisting,
+                                              targetExists));
+                    }
+                }
+            }
+        }
+
+        return list.stream();
+    }
+
+    @ParameterizedTest
+    @MethodSource("params")
+    void copyFollow(PathType type, String mode, boolean replaceExisting,
+                    boolean targetExists) throws IOException {
+        op(OpType.COPY, type, mode, replaceExisting, targetExists, true);
+    }
+
+    @ParameterizedTest
+    @MethodSource("params")
+    void copyNoFollow(PathType type, String mode, boolean replaceExisting,
+                    boolean targetExists) throws IOException {
+        op(OpType.COPY, type, mode, replaceExisting, targetExists, false);
+    }
+
+    @ParameterizedTest
+    @MethodSource("params")
+    void move(PathType type, String mode, boolean replaceExisting,
+              boolean targetExists) throws IOException {
+        op(OpType.MOVE, type, mode, replaceExisting, targetExists, false);
+    }
+
+    void op(OpType op, PathType type, String mode, boolean replaceExisting,
+            boolean targetExists, boolean followLinks) throws IOException {
+
+        Path source = null;
+        Path target = null;
+        Path linkTarget = null;
+        try {
+            switch (type) {
+                case FILE ->
+                    source = Files.createTempFile("file", "dat");
+                case DIR ->
+                    source = Files.createTempDirectory("dir");
+                case LINK -> {
+                    linkTarget = Files.createTempFile("link", "target");
+                    Path link = Path.of("link");
+                    source = Files.createSymbolicLink(link, linkTarget);
+                }
+            }
+
+            if (Files.getFileStore(source).supportsFileAttributeView("posix")) {
+                Set<PosixFilePermission> perms =
+                    PosixFilePermissions.fromString(mode);
+                if (op == OpType.COPY && type == PathType.LINK && followLinks)
+                    Files.setPosixFilePermissions(linkTarget, perms);
+                else
+                    Files.setPosixFilePermissions(source, perms);
```

Comment:
Should this be expanded to cover the permissions of the directory containing the source file, that's the case where the access(2) call may fail because the directory can't be updated.

Reply:
From what I can tell, `access(source, R_OK)` will fail if the parent directory does not confer adequate permissions to read the file. I would think it would be the parent directory of the target which would need to be checked.